File: Tools/protocol-probe/messages.py
"""
WHOOP 5.0 / MG message catalogue.

Only decodes what is *publicly documented* by the NOOP project and its
sources (goose, dofek, judes.club). Fields NOT listed here are unknown --
`decode_puffin_payload` returns an UnknownMessage with the raw bytes rather
than guessing, so you don't get silently wrong output.

BLE GATT layout (from NOOP docs):
    0x2A37              standard Heart Rate Measurement char (strap -> app,
                         no bonding required)
    fd4b0002            app -> strap, write 0xAA-framed commands
    fd4b0003/4/5/7      strap -> app, notify 0xAA-framed responses/data/console
"""

from __future__ import annotations

import struct
from dataclasses import dataclass, field as dc_field
from typing import Optional

from .framing import InnerMessage

# --- Known command opcodes (inner `cmd` byte), from Asherlc/dofek + goose ---
CMD_START_DEVICE_CONFIG_KEY_EXCHANGE = 0x73
CMD_SET_FF_VALUE = 0x78          # SET_CONFIG -- one feature-flag write
CMD_GET_HELLO = 0x01              # b3 = 0x01
CMD_GET_DATA_RANGE = 0x02         # b3 = 0x00
CMD_SEND_HISTORICAL_DATA = 0x03   # b3 = 0x00

COMMAND_NAMES = {
    CMD_START_DEVICE_CONFIG_KEY_EXCHANGE: "START_DEVICE_CONFIG_KEY_EXCHANGE",
    CMD_SET_FF_VALUE: "SET_FF_VALUE (SET_CONFIG)",
    CMD_GET_HELLO: "GET_HELLO",
    CMD_GET_DATA_RANGE: "GET_DATA_RANGE",
    CMD_SEND_HISTORICAL_DATA: "SEND_HISTORICAL_DATA",
}

# Inner envelope type byte seen for the puffin command/response wrapper.
PUFFIN_ENVELOPE_TYPE = 0x23

# Historical/streamed record type byte (first byte of payload on the data
# channels), per NOOP's WHOOP5_DEEP_DATA.md -- the "R22" deep biometric
# packet, gated behind the enable_r22_packets feature flag.
RECORD_TYPE_DEEP_BIOMETRIC = 0x2F


@dataclass
class FeatureFlagWrite:
    """A decoded SET_FF_VALUE (cmd 0x78) command body."""
    name: str
    value_char: str
    raw: bytes

    @classmethod
    def decode(cls, payload: bytes) -> "FeatureFlagWrite":
        # 40-byte body: b3(1, already stripped by caller) is separate;
        # payload here = 32-byte ASCII NUL-padded name + 1 value byte + 7 zero pad
        if len(payload) < 33:
            raise ValueError(f"SET_FF_VALUE payload too short: {len(payload)}")
        name_raw = payload[:32]
        name = name_raw.split(b"\x00", 1)[0].decode("ascii", errors="replace")
        value_char = chr(payload[32]) if payload[32] else ""
        return cls(name=name, value_char=value_char, raw=payload)


@dataclass
class DeepBiometricRecord:
    """
    Partial decode of a type-0x2F record.

    Only the two fields the community has actually confirmed are decoded:
      - heart_rate_bpm at payload byte 14 (uint8)
      - accel x/y/z as little-endian float32 at bytes 37/41/45

    Everything else in the record is preserved in `raw` and in `unknown_span`
    so you can diff/annotate it yourself as more of the format gets mapped --
    this is explicitly still an open reverse-engineering question upstream
    (see NOOP issue #174), so treat bytes outside these two fields as opaque.
    """
    heart_rate_bpm: Optional[int]
    accel_x: Optional[float]
    accel_y: Optional[float]
    accel_z: Optional[float]
    raw: bytes

    @classmethod
    def decode(cls, payload: bytes) -> "DeepBiometricRecord":
        hr = payload[14] if len(payload) > 14 else None
        ax = ay = az = None
        if len(payload) >= 49:
            ax, ay, az = struct.unpack_from("<fff", payload, 37)
        return cls(heart_rate_bpm=hr, accel_x=ax, accel_y=ay, accel_z=az, raw=payload)


@dataclass
class UnknownMessage:
    envelope_type: int
    cmd: int
    b3: int
    raw_payload: bytes


DecodedPuffin = object  # FeatureFlagWrite | DeepBiometricRecord | UnknownMessage


def decode_puffin_payload(inner: InnerMessage) -> DecodedPuffin:
    """Dispatch an already-frame-parsed InnerMessage to a field decoder."""
    if inner.type != PUFFIN_ENVELOPE_TYPE:
        return UnknownMessage(inner.type, inner.cmd, inner.b3, inner.payload)

    if inner.cmd == CMD_SET_FF_VALUE:
        try:
            return FeatureFlagWrite.decode(inner.payload)
        except ValueError:
            pass

    if inner.payload and inner.payload[0] == RECORD_TYPE_DEEP_BIOMETRIC:
        return DeepBiometricRecord.decode(inner.payload[1:])

    return UnknownMessage(inner.type, inner.cmd, inner.b3, inner.payload)


# --- Standard BLE Heart Rate Measurement (0x2A37) -- this part is fully
# standardized by the Bluetooth SIG, not WHOOP-specific, and needs no
# reverse engineering. ---

@dataclass
class HeartRateMeasurement:
    bpm: int
    sensor_contact_supported: bool
    sensor_contact_detected: bool
    energy_expended_kj: Optional[int]
    rr_intervals_s: list

    @classmethod
    def decode(cls, data: bytes) -> "HeartRateMeasurement":
        if not data:
            raise ValueError("empty HR measurement")
        flags = data[0]
        hr_16bit = bool(flags & 0x01)
        contact_supported = bool(flags & 0x04)
        contact_detected = bool(flags & 0x02)
        energy_present = bool(flags & 0x08)
        rr_present = bool(flags & 0x10)

        idx = 1
        if hr_16bit:
            bpm = struct.unpack_from("<H", data, idx)[0]
            idx += 2
        else:
            bpm = data[idx]
            idx += 1

        energy = None
        if energy_present:
            energy = struct.unpack_from("<H", data, idx)[0]
            idx += 2

        rr_intervals = []
        if rr_present:
            while idx + 2 <= len(data):
                raw_rr = struct.unpack_from("<H", data, idx)[0]
                rr_intervals.append(raw_rr / 1024.0)  # spec: 1/1024 s units
                idx += 2

        return cls(bpm=bpm, sensor_contact_supported=contact_supported,
                    sensor_contact_detected=contact_detected,
                    energy_expended_kj=energy, rr_intervals_s=rr_intervals)

File: Packages/WhoopKitExperimental/Sources/WhoopKit/Messages.swift
//
//  Messages.swift
//  WhoopKit
//
//  Same scope note as the Python version: only decodes what's publicly
//  documented (NOOP + goose + dofek + judes.club). Unknown frames come back
//  as .unknown with raw bytes rather than a guessed decode.
//

import Foundation

public enum Command {
    public static let startDeviceConfigKeyExchange: UInt8 = 0x73
    public static let setFFValue: UInt8 = 0x78          // SET_CONFIG
    public static let getHello: UInt8 = 0x01              // b3 = 0x01
    public static let getDataRange: UInt8 = 0x02          // b3 = 0x00
    public static let sendHistoricalData: UInt8 = 0x03    // b3 = 0x00
}

public let puffinEnvelopeType: UInt8 = 0x23
public let recordTypeDeepBiometric: UInt8 = 0x2F

// The 15 feature flags NOOP sends on the opt-in "unlock deep data" path.
// Values are the ASCII characters the strap expects; see NOOP's
// Whoop5Config.swift for the golden-tested canonical list -- reproduce that
// exact list from your own capture/NOOP checkout before relying on it, this
// is deliberately left as a name-only stub so you don't ship unverified
// values against your own hardware.
public let knownFeatureFlagNames: [String] = [
    "enable_r22_packets",
    // ... remaining 14 flags: pull from your own verified capture or from
    // NOOP's Whoop5Config.swift rather than trusting a hardcoded guess here.
]

public struct FeatureFlagWrite {
    public let name: String
    public let valueChar: Character?
    public let raw: Data

    public static func decode(_ payload: Data) -> FeatureFlagWrite? {
        guard payload.count >= 33 else { return nil }
        let bytes = [UInt8](payload)
        let nameBytes = bytes[0..<32]
        let nameEnd = nameBytes.firstIndex(of: 0) ?? 32
        let name = String(bytes: nameBytes[0..<nameEnd], encoding: .ascii) ?? "<invalid>"
        let valueByte = bytes[32]
        let valueChar = valueByte != 0 ? Character(UnicodeScalar(valueByte)) : nil
        return FeatureFlagWrite(name: name, valueChar: valueChar, raw: payload)
    }
}

public struct DeepBiometricRecord {
    /// byte 14 of the record payload -- confirmed by community captures.
    public let heartRateBPM: UInt8?
    /// little-endian float32 at bytes 37/41/45 -- confirmed by community captures.
    public let accelX: Float?
    public let accelY: Float?
    public let accelZ: Float?
    public let raw: Data

    public static func decode(_ payload: Data) -> DeepBiometricRecord {
        let bytes = [UInt8](payload)
        let hr: UInt8? = bytes.count > 14 ? bytes[14] : nil

        var ax: Float?, ay: Float?, az: Float?
        if bytes.count >= 49 {
            ax = readFloatLE(bytes, at: 37)
            ay = readFloatLE(bytes, at: 41)
            az = readFloatLE(bytes, at: 45)
        }
        return DeepBiometricRecord(heartRateBPM: hr, accelX: ax, accelY: ay, accelZ: az, raw: payload)
    }

    private static func readFloatLE(_ bytes: [UInt8], at offset: Int) -> Float {
        let slice = bytes[offset..<offset + 4]
        let bits = UInt32(slice[slice.startIndex]) |
            (UInt32(slice[slice.startIndex + 1]) << 8) |
            (UInt32(slice[slice.startIndex + 2]) << 16) |
            (UInt32(slice[slice.startIndex + 3]) << 24)
        return Float(bitPattern: bits)
    }
}

public enum DecodedPuffin {
    case featureFlag(FeatureFlagWrite)
    case deepBiometric(DeepBiometricRecord)
    case unknown(type: UInt8, cmd: UInt8, b3: UInt8, payload: Data)
}

public func decodePuffinPayload(_ inner: InnerMessage) -> DecodedPuffin {
    guard inner.type == puffinEnvelopeType else {
        return .unknown(type: inner.type, cmd: inner.cmd, b3: inner.b3, payload: inner.payload)
    }
    if inner.cmd == Command.setFFValue, let ff = FeatureFlagWrite.decode(inner.payload) {
        return .featureFlag(ff)
    }
    if let first = inner.payload.first, first == recordTypeDeepBiometric {
        return .deepBiometric(DeepBiometricRecord.decode(inner.payload.dropFirst()))
    }
    return .unknown(type: inner.type, cmd: inner.cmd, b3: inner.b3, payload: inner.payload)
}

// --- Standard Bluetooth SIG Heart Rate Measurement (0x2A37) --------------
// Fully standardized, not WHOOP-specific.

public struct HeartRateMeasurement {
    public let bpm: Int
    public let sensorContactSupported: Bool
    public let sensorContactDetected: Bool
    public let energyExpendedKJ: Int?
    public let rrIntervalsSeconds: [Double]

    public static func decode(_ data: Data) -> HeartRateMeasurement? {
        guard !data.isEmpty else { return nil }
        let bytes = [UInt8](data)
        let flags = bytes[0]
        let hr16 = flags & 0x01 != 0
        let contactSupported = flags & 0x04 != 0
        let contactDetected = flags & 0x02 != 0
        let energyPresent = flags & 0x08 != 0
        let rrPresent = flags & 0x10 != 0

        var idx = 1
        let bpm: Int
        if hr16 {
            guard bytes.count >= idx + 2 else { return nil }
            bpm = Int(bytes[idx]) | (Int(bytes[idx + 1]) << 8)
            idx += 2
        } else {
            guard bytes.count > idx else { return nil }
            bpm = Int(bytes[idx])
            idx += 1
        }

        var energy: Int?
        if energyPresent {
            guard bytes.count >= idx + 2 else { return nil }
            energy = Int(bytes[idx]) | (Int(bytes[idx + 1]) << 8)
            idx += 2
        }

        var rr: [Double] = []
        if rrPresent {
            while idx + 2 <= bytes.count {
                let raw = Int(bytes[idx]) | (Int(bytes[idx + 1]) << 8)
                rr.append(Double(raw) / 1024.0)
                idx += 2
            }
        }

        return HeartRateMeasurement(bpm: bpm, sensorContactSupported: contactSupported,
                                     sensorContactDetected: contactDetected,
                                     energyExpendedKJ: energy, rrIntervalsSeconds: rr)
    }
}
